File: Data Structures & Algorithms/three-integer-sum/submission-9.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        for i, num in enumerate(nums):
            if num > 0:
                break
            if i-1 >= 0 and nums[i] == nums[i-1]:
                continue
            
            l, r = i+1, len(nums) - 1
            while l < r:
                if num + nums[l] + nums[r] == 0:
                    res.append([num, nums[l], nums[r]])
                    l += 1
                    r -= 1
                    while l < r and nums[l-1] == nums[l]:
                        l += 1
                elif nums[i] + nums[l] + nums[r] < 0:
                    l += 1
                else:
                    r -= 1
            
        return res

File: Data Structures & Algorithms/top-k-elements-in-list/submission-9.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> counts;
        vector<vector<int>> freq(nums.size() + 1);

        for (int num : nums) {
            counts[num] += 1;
        }

        for (auto& pair : counts) {
            freq[pair.second].push_back(pair.first);
        }

        vector<int> res;

        for (int i = freq.size() - 1; k > 0; --i) {
            for(int j = 0; j < freq[i].size(); ++j) {
                res.push_back(freq[i][j]);
                --k;
            }
        }

        return res;
    }
};
